kornia/feature/mkd.py

```diff
@@ -0,0 +1,609 @@
+                                         reduce_dims=128)
+        >>> desc = mkd(patches) # 23x128
+    """
+
+    def __init__(self,
+                 patch_size: int = 32,
+                 dtype: str = 'concat',
+                 whitening: str = 'pcawt',
+                 training_set: str = 'liberty',
+                 reduce_dims: int = 128) -> None:
+        super().__init__()
+
+        self.patch_size = patch_size
+        self.dtype = dtype
+        self.whitening = whitening
+        self.training_set = training_set
+
+        self.sigma = 1.4 * (patch_size / 64)
+        self.smoothing = GaussianBlur2d((5, 5),
+                                        (self.sigma, self.sigma),
+                                        'replicate')
+        self.gradients = MKDGradients()
+
+        # Cartesian embedding with absolute gradients.
+        if dtype in ['cart', 'concat']:
+            ori_abs = EmbedGradients(patch_size=patch_size,
+                                     relative=False)
+            cart_emb = ExplicitSpacialEncoding(dtype='cart',
+                                               fmap_size=patch_size,
+                                               in_dims=ori_abs.kernel.d)
+            self.cart_feats = nn.Sequential(ori_abs, cart_emb)
+
+        # Polar embedding with relative gradients.
+        if dtype in ['polar', 'concat']:
+            ori_rel = EmbedGradients(patch_size=patch_size,
+                                     relative=True)
+            polar_emb = ExplicitSpacialEncoding(dtype='polar',
+                                                fmap_size=patch_size,
+                                                in_dims=ori_rel.kernel.d)
+            self.polar_feats = nn.Sequential(ori_rel, polar_emb)
+
+        self.odims: int = 0
+        if dtype == 'concat':
+            self.odims = polar_emb.odims + cart_emb.odims
+        elif dtype == 'cart':
+            self.odims = cart_emb.odims
+        elif dtype == 'polar':
+            self.odims = polar_emb.odims
+
+        # Compute true reduce_dims.
+        self.reduce_dims: int = min(reduce_dims, self.odims)
+
+        # Load supervised(lw)/unsupervised(pca) model trained on training_set.
+        if self.whitening is not None:
+            whitening_models = torch.hub.load_state_dict_from_url(
+                urls[self.dtype], map_location=lambda storage, loc: storage
+            )
+            whitening_model = whitening_models[training_set]
+            self.whitening_layer = Whitening(whitening,
+                                             whitening_model,
+                                             in_dims=self.odims,
+                                             reduce_dims=self.reduce_dims)
+            self.odims = self.reduce_dims
+
+    def forward(self, patches: torch.Tensor) -> torch.Tensor:
+        # Extract gradients.
+        g = self.smoothing(patches)
+        g = self.gradients(g)
+
+        # Extract polar and/or cart features.
```

Comment:
I don't get how is this used. Can the user pass both polar/cart and/or concat? Can this better be resolved inside the constructor. To much ifelse branching here and below

Reply:
Basially, for polar and cart, the model looks like:
`input -> smooth -> grads -> Position embedding -> Whiten`
but in the concat case,
```
input -> smooth -> grads -> pos1 -\
                           - pos2 -> concat -> Whiten
```
So it breaks the sequential chain.
I can try to refactor to remove branching by using dicts instead.

Reply:
great